src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -1188,6 +1195,31 @@ public override int GetValues(object[] values)
         /// <returns>The value of the specified column.</returns>
         public NpgsqlDateTime GetTimeStamp(int ordinal) => GetFieldValue<NpgsqlDateTime>(ordinal);
 
+        /// <inheritdoc />
+        protected override DbDataReader GetDbDataReader(int ordinal)
+        {
+            var field = CheckRowAndGetField(ordinal);
+            var type = field.PostgresType;
+            var isArray = type is PostgresArrayType;
+            var elementType = isArray ? ((PostgresArrayType)type).Element : type;
+            if (elementType.InternalName != "record" && !(elementType is PostgresCompositeType))
+                throw new InvalidCastException("GetDbDataReader() not supported for type " + field.TypeDisplayName);
+
+            SeekToColumn(ordinal, false).GetAwaiter().GetResult();
+            if (ColumnLen == -1)
+                ThrowHelper.ThrowInvalidCastException_NoValue(field);
+
+            if (_isSequential)
+                throw new InvalidOperationException("GetDbDataReader not supported in sequential mode.");
+
+            var reader = new NpgsqlRecordDataReader(this, UniqueRowId, 1);
```

Comment:
To reduce allocations, this can be recycled. In other words, if the user disposes an NpgsqlRecordDataReader before requesting another, we can hand out the same instance.

Reply:
Yes, am planning for this. :)
Any good idea how to best implement this? Like a cache. In case of the user requests multiple levels and multiple columns, each having an inner data reader.

Reply:
Yeah, each reader (both the top-level and every inner one) could have a `_cachedInnerDataReader` which is initially null. When we need to hand out an NpgsqlInnerDataReader, if it's null we allocate a new one and return it. When NpgsqlInnerDataReader is disposed, we set `_cachedInnerDataReader` to point to it.